clang/include/clang/AST/PreorderAST.h

```diff
@@ -0,0 +1,174 @@
+      }
+
+      void addVar(std::string name) {
+        // We initialize the count of each variable in a node to 1.
+        variables.push_back(VarTy {name, getConstVal(1)});
+      }
+
+      // Is the operator commutative and associative?
+      bool isOpCommutativeAndAssociative() {
+        return opcode == BO_Add ||
+               opcode == BO_Mul;
+      }
+    };
+
+  private:
+    ASTContext &Ctx;
+    Lexicographic Lex;
+    llvm::raw_ostream &OS;
+    bool HasError;
+    ASTNode *AST;
+
+  public:
+    PreorderAST(ASTContext &Ctx, Expr *E) :
+      Ctx(Ctx), Lex(Lexicographic(Ctx, nullptr)),
+      OS(llvm::outs()), HasError(false) {
+
+      HasError = false;
+      AST = new ASTNode(Ctx);
+      insert(AST, E);
+      normalize(AST, HasError);
+    }
+
+    // Create a preorder AST from the expression E.
+    // @param[in] N is the current node of the AST.
+    // @param[in] E is the sub expression which needs to be added to N.
+    // @param[in] Parent is the parent node for N.
+    void insert(ASTNode *N, Expr *E, ASTNode *Parent = nullptr);
+
+    // Normalize the input expression through a series of transforms on the
+    // preorder AST.
+    // @param[in] N is the root of the AST.
+    // @param[out] HasError is populated if an error is encountered during
+    // normalization.
+    void normalize(ASTNode *N, bool &HasError);
+
+    // Sort the variables in a node of the AST.
+    // @param[in] N is the root of the AST.
+    // @param[out] HasError is populated if an error is encountered during
+    // sorting.
+    void sort(ASTNode *N, bool &HasError);
+
+    // Check if the two ASTs N1 and N2 are equal.
+    // @param[in] N1 is the first AST.
+    // @param[in] N2 is the second AST.
+    // @return Returns a boolean indicating whether N1 and N2 are equal.
+    bool isEqual(ASTNode *N1, ASTNode *N2);
+
+    // Compare the current AST with the given AST. This in turn, invokes
+    // isEqual(N1, N2);
+    // @param[in] this is the first AST.
+    // @param[in] PT is the second AST.
+    // @return Returns a value of type Lexicographic::Result indicating whether
+    // the two ASTs are equal or not. 
+    Result compare(PreorderAST &PT);
+
+    // Cleanup the memory consumed by the AST.
+    // @param[in] N is the root node of the AST.
+    void cleanup(ASTNode *N);
+
+    // Cleanup the memory consumed by the AST. This is intended to be called
+    // from outside the PreorderAST class.
+    void cleanup();
+
+    // Check if an error has occurred during normalization of the expression.
+    // @return Whether an error has occurred or not.
+    bool hasError() {
+      return HasError;
+    }
+
+    // Print the preorder AST.
+    // @param[in] N is the root node of the AST.
+    void print(ASTNode *N);
+
+    // Invoke IgnoreValuePreservingOperations to strip off casts.
+    // @param[in] E is the expression whose casts must be stripped.
+    // @return E with casts stripped off.
+    Expr *IgnoreCasts(const Expr *E);
```

Comment:
The name IgnoreCasts is somewhat confusing since Lex.IgnoreValuePreservingOperations also ignores non-cast operations such as `&*e`.